tsl/src/chunk_copy.c

```diff
@@ -0,0 +1,581 @@
+
+struct ChunkCopyStage
+{
+	const char *name;
+	chunk_copy_stage_func function;
+	/* todo: abort function */
+};
+
+/* To track a chunk move or copy activity */
+struct ChunkCopy
+{
+	/* catalog data */
+	FormData_chunk_copy_activity fd;
+	/* current stage being executed */
+	const ChunkCopyStage *stage;
+	/* chunk to copy */
+	Chunk *chunk;
+	/* from/to foreign servers */
+	ForeignServer *src_server;
+	ForeignServer *dst_server;
+	/* temporary memory context */
+	MemoryContext mcxt;
+};
+
+static HeapTuple
+chunk_copy_activity_make_tuple(const FormData_chunk_copy_activity *fd, TupleDesc desc)
+{
+	Datum values[Natts_chunk_copy_activity];
+	bool nulls[Natts_chunk_copy_activity] = { false };
+	memset(values, 0, sizeof(values));
+	values[AttrNumberGetAttrOffset(Anum_chunk_copy_activity_id)] = Int32GetDatum(fd->id);
+	values[AttrNumberGetAttrOffset(Anum_chunk_copy_activity_operation_id)] =
+		NameGetDatum(&fd->operation_id);
+	values[AttrNumberGetAttrOffset(Anum_chunk_copy_activity_backend_pid)] =
+		Int32GetDatum(fd->backend_pid);
+	values[AttrNumberGetAttrOffset(Anum_chunk_copy_activity_completed_stage)] =
+		NameGetDatum(&fd->completed_stage);
+	values[AttrNumberGetAttrOffset(Anum_chunk_copy_activity_time_start)] =
+		TimestampTzGetDatum(fd->time_start);
+	values[AttrNumberGetAttrOffset(Anum_chunk_copy_activity_chunk_id)] =
+		Int32GetDatum(fd->chunk_id);
+	values[AttrNumberGetAttrOffset(Anum_chunk_copy_activity_source_node_name)] =
+		NameGetDatum(&fd->source_node_name);
+	values[AttrNumberGetAttrOffset(Anum_chunk_copy_activity_dest_node_name)] =
+		NameGetDatum(&fd->dest_node_name);
+	values[AttrNumberGetAttrOffset(Anum_chunk_copy_activity_delete_on_src_node)] =
+		BoolGetDatum(fd->delete_on_src_node);
+	return heap_form_tuple(desc, values, nulls);
+}
+
+static void
+chunk_copy_activity_insert_rel(Relation rel, const FormData_chunk_copy_activity *fd)
+{
+	CatalogSecurityContext sec_ctx;
+	HeapTuple new_tuple;
+
+	new_tuple = chunk_copy_activity_make_tuple(fd, RelationGetDescr(rel));
+
+	ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+	ts_catalog_insert(rel, new_tuple);
+	ts_catalog_restore_user(&sec_ctx);
+	heap_freetuple(new_tuple);
+}
+
+static void
+chunk_copy_activity_insert(const FormData_chunk_copy_activity *fd)
+{
+	Catalog *catalog;
+	Relation rel;
+
+	catalog = ts_catalog_get();
+	rel = table_open(catalog_get_table_id(catalog, CHUNK_COPY_ACTIVITY), RowExclusiveLock);
+
+	chunk_copy_activity_insert_rel(rel, fd);
+	table_close(rel, RowExclusiveLock);
+}
+
+static ScanTupleResult
+chunk_copy_activity_tuple_update(TupleInfo *ti, void *data)
+{
+	ChunkCopy *cc = data;
+	Datum values[Natts_chunk_copy_activity];
+	bool nulls[Natts_chunk_copy_activity];
+	CatalogSecurityContext sec_ctx;
+	bool should_free;
+	HeapTuple tuple = ts_scanner_fetch_heap_tuple(ti, false, &should_free);
+	HeapTuple new_tuple;
+
+	heap_deform_tuple(tuple, ts_scanner_get_tupledesc(ti), values, nulls);
```

Comment:
This is completely optional, but now it is possible to copy the tuple and update it directly with GETSTRUCT to get a pointer to the FormData. This might simplify the code a bit, but this deform and form works too.

Reply:
But it still requires to fill values in in order to update tuple, not sure that it is more convenient here to use